### DISCUSSION TITLE
Connecting IGT tracking Device with OpenIGTLink to Slicer  by PLUS

### DISCUSSION THREAD

**Discussion by Barneyhuuu** [PlusToolkit/PlusLib#971] [Q&A] [answered]
Hi!
We're using a new optical locator ( AimPosition ), which is OpenIGTLink compatible. We overwrote the sample file "trackerServer. CXX" in the source code to transfer real-time location information through OpenIGTLinkIF module to the 3D Slicer. But rather than we expected, Plus was not involved in the process. this means we can't use PLUS to complete spatial and temporal calibration.
So, is there something wrong?
<img width="1412" alt="2022-07-27 6 57 19" src="https://user-images.githubusercontent.com/93598282/181231106-e11ba29c-7b5e-4379-8248-1e53ebf3d947.png">

**Comment by Sunderlandkyl** [comment]
What ports are you using for the input/output of Plus?

The output port of trackerServer should match the input port of Plus, and should be different from the output port of Plus (which should match Slicer's input port).

**Reply by Barneyhuuu** [reply]
Thank you! I'm still a little confused, the tracking device is connected to the same computer via USB, so the SeverAddress is also the   computer's IP address?  And the output port of trackServer is also Server port? How do I set output port of Plus in the config file?
Again, thank you very much, you are being very helpful !
<img width="443" alt="2022-07-27 10 59 45" src="https://user-images.githubusercontent.com/93598282/181287679-96decc3d-3082-41a3-8b09-b1eeb8176ef9.png">

!

**Reply by Sunderlandkyl** [reply]
The PlusOpenIGTLinkServer element in the config file controls the output parameters of Plus. The ListeningPort attribute defines the output port. There is documentation on PlusOpenIGTLinkServer here: http://perk-software.cs.queensu.ca/plus/doc/nightly/user/ApplicationPlusServer.html.

```
<PlusOpenIGTLinkServer
  MaxNumberOfIgtlMessagesToSend="10"
  MaxTimeSpentWithProcessingMs="50"
  ListeningPort="18944"
  MissingInputGracePeriodSec="5.0"
  SendValidTransformsOnly="TRUE"
  OutputChannelId="SomeValidChannel" >
  <DefaultClientInfo>
    <MessageTypes>
      <Message Type="IMAGE" />
      <Message Type="TRANSFORM" />
    </MessageTypes>
    <TransformNames>
      <Transform Name="StylusTipToReference" />
      <Transform Name="ProbeToReference" />
    </TransformNames>
  </DefaultClientInfo>
</PlusOpenIGTLinkServer>
```

**Reply by Barneyhuuu** [reply]
Dear Mr Sunderland,
    Thank you for your help! we sucessfully sending real-time location information through OpenIGTLinkIF module directly to the 3D Slicer.
Then we overwrote the sample file in this wedsite:http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceOpenIGTLinkTracker.html. But, it seemed that Plus did not receive any information.
![92153cf9c9b929521af10667a519194](https://user-images.githubusercontent.com/93598282/183085957-b76f4490-487a-49a4-9d27-ba7dfeef1ab0.png)
I don't know what went wrong, could you please help me？Thank you!
OpenIGTlink interface .cpp file: [（trackerServer）TransTooltoReference.txt](https://github.com/PlusToolkit/PlusLib/files/9269427/trackerServer.TransTooltoReference.txt)
PLIUS config file: 
[PlusDeviceSet_Server_Aimposition.txt](https://github.com/PlusToolkit/PlusLib/files/9269443/PlusDeviceSet_Server_Aimposition.txt)

**Reply by Sunderlandkyl** [reply]
Can you try changing TrackerInternalCoordinateSystemName to "Tracker"?

Are you intending to incorporate the video stream from SavedDataSource?

Can you also upload a full log file?

**Reply by Barneyhuuu** [reply]
Dear Mr Sunderland,
Thank you for your help! I changed the TrackerInternalCoordinateSystemName to "Tracker" but unfortunately it didn't work. We can see the image is correctly imported, but there is no transform information, the transforms were sent out.
![1](https://user-images.githubusercontent.com/93598282/183808511-d2cf7a2e-2da5-47e6-b26c-e2f8d0c5e4c0.PNG)
I uploaded the full log file and  the modified PLUS config file, OpenIGTLink track server cpp file if you need.
Thank you again!
[File.zip](https://github.com/PlusToolkit/PlusLib/files/9296417/File.zip)

**Comment by Sunderlandkyl** [comment, marked as answer]
Sorry, I was preoccupied with a lecture and didn't have time to take another look until now.

The transform names that are sent from your C++ class "ReferenceStar", "Pointer", "Probe" work if your message type is TDATA.
Because you are sending your transforms as igtl::TransformMessage, the transform names must specify the coordinate systems that they are transforming between. "ReferenceStarToTracker", "PointerToTracker", "ProbeToTracker".

Due to the length of "ReferenceStarToTracker", you may also have to rename it to something shorter (ex. "ReferenceToTracker"). Device names in OpenIGTLink can only be up to 20 characters long.

**Comment by clllng** [comment]
Dear Sunderland,   @Sunderlandkyl 
Hi!
We also have a very similar problem using Plus for data transmission which the tracking device is OpenIGTLink compatible. Followed the example “TooltoReference” in this page, tracking data could be sent to 3D Slicer successfully. But we had trouble when we tried to use Plus to collect the tracking data. The same errors occurred as Barney reported above "No data is broadcasted , as no data is available".
[PlusLog 1.txt](https://github.com/PlusToolkit/PlusLib/files/10385595/PlusLog.1.txt)
So, We changed the device names to “RenferencetToTracker”, ”PointerToTracker” and “ProbeToTracker”as you kindly suggested above, and also unified the Data Source IDs in configuration file. [Plus config  2.txt](https://github.com/PlusToolkit/PlusLib/files/10385600/Plus.config.2.txt)
But Plus reported the error of “Unable to parse transform name, there are 2 matching 'To' phrases in the transform name 'ProbeToTrackerToTracker', while exactly one allowed.[PlusLog 3.txt](https://github.com/PlusToolkit/PlusLib/files/10385611/PlusLog.3.txt)
We looked up every configuration example on users guide and we don’t know what to do next, could you please help? Thank you!
Kind regards,
Cling

**Reply by Sunderlandkyl** [reply]
I think that DataSource ids in the device should not contain "ToTracker".

Ex.
```
	  <DataSources>
			<DataSource Type="Tool" Id="Reference" />
			<DataSource Type="Tool" Id="Pointer" />
			<DataSource Type="Tool" Id="Probe" />
		</DataSources>
		<OutputChannels>
			<OutputChannel Id="TrackerStream">
				<DataSource Id="Reference" />
				<DataSource Id="Pointer" />
				<DataSource Id="Probe" />
			</OutputChannel>
		</OutputChannels>
```

**Reply by clllng** [reply]
Dear Mr Sunderland,
Thank you for your help! I tried changing the data source id as your suggestion. But unfortunately, The error still occurred ,"No data is broadcasted , as no data is available" [PlusLog 4.txt](https://github.com/PlusToolkit/PlusLib/files/10395661/PlusLog.4.txt)
Is there any problem with TransformNames?  3D slicers still cannot receive the transforms . Could you please help me ? Thank you ! 
It's our config files : 
[Plus config  3.txt](https://github.com/PlusToolkit/PlusLib/files/10395697/Plus.config.3.txt)

**Reply by ScalpelCode** [reply]
Hello, have you solved this problem? I also encountered the same problem.
